### DISCUSSION TITLE
Trouble Configuring Xdebug in Official FrankenPHP Docker Image

### DISCUSSION THREAD

**Discussion by adelynx** [php/frankenphp#492] [Q&A] [answered]
**Description:**

Hello FrankenPHP community!

I'm currently using the official FrankenPHP Docker image for my project, and I'm having some trouble configuring Xdebug to work seamlessly within this environment. I've followed the usual steps for Xdebug configuration, but it doesn't seem to be working as expected.

Here are the steps I've taken so far:

1. Installed Xdebug extension.
2. Updated php.ini with appropriate Xdebug settings.
3. Restarted the PHP service.

Despite these steps, I'm unable to connect my IDE (e.g., VSCode) to Xdebug within the FrankenPHP Docker container.

**My environment details:**
- Official FrankenPHP Docker image version: latest
- Operating System: Ubuntu 23.10
- PHP Version: 8.3.2-1
- IDE: PhpStorm

**Questions:**
1. Has anyone successfully configured Xdebug within the official FrankenPHP Docker image?
2. Are there any specific considerations or steps that need to be taken in FrankenPHP Docker environment?
3. Is there a recommended Xdebug configuration for FrankenPHP?

I would greatly appreciate any guidance or advice from the community on how to make Xdebug work seamlessly within the FrankenPHP Docker image.

Thank you for your help! 🙏

**Comment by adelynx** [comment, marked as answer]
# Solved

The [docker-php-extension-installer](https://github.com/mlocati/docker-php-extension-installer) script is provided in the base image. Adding additional PHP extensions is straightforward:

```docker
FROM dunglas/frankenphp

# add additional extensions here:
RUN install-php-extensions \
    pdo_mysql \
    gd \
    intl \
    zip \
    opcache
```

**Reply by withinboredom** [reply]
This is from a real prod service at work:

```dockerfile
FROM dunglas/frankenphp:1-php8.3-bookworm

ARG EXTRA_EXTENSIONS="opcache"
RUN install-php-extensions @composer intl ctype mbstring apcu pdo pdo_pgsql $EXTRA_EXTENSIONS

RUN cp "$PHP_INI_DIR/php.ini-production" "$PHP_INI_DIR/php.ini"; \
    sed -i 's/variables_order = "GPCS"/variables_order = "EGPCS"/' "$PHP_INI_DIR/php.ini"; \
    echo "\nxdebug.mode=debug\nxdebug.discover_client_host=1\n" >> $PHP_INI_DIR/php.ini;

ENV SERVER_NAME=":8080" FRANKENPHP_CONFIG="worker /app/public/index.php 32"

COPY composer.json composer.lock /app/

# only run install when dependencies change
RUN composer install --no-dev --optimize-autoloader --apcu-autoloader --no-scripts --no-plugins

COPY . /app/

RUN composer dump --optimize --apcu --no-dev
```

Then in the docker compose file:


```yaml
version: "3.8"

x-service-build: &build
  context: .
  pull: true
  dockerfile: Dockerfile
  target: production
  # enable step debugging:
  args:
    EXTRA_EXTENSIONS: "xdebug"

x-service-env: &env
  # enable step debugging:
  FRANKENPHP_CONFIG: ""

services:
  a-service:
    build:
      <<: *build
    environment:
      <<: *env
```

This is for a microservice, so there's no symfony or anything like that (just slim + doctrine + a few other things). Our worker script looks something like:

```php
require_once __DIR__ . '/../vendor/autoload.php';
$container = require __DIR__ . '/../src/bootstrap.php';

// prevent users cancelling a request from crashing a worker
ignore_user_abort(true);

function handle_request() {
    global $container;

    // set the default to a 404 instead of caddy's 200
    http_response_code(404);

    try {
        $container->get(App::class)->run();
    } catch (Throwable $exception) {
        // handle any uncaught exceptions and rethrow for frankenphp to output to logs
        http_response_code(500);

        // and make sure we reset the entity manager so the next request has a clean slate
        $container->get(EntityManager::class)->clear();

        throw $exception;
    }
}

// if we are not in worker mode, we can just handle a single request and die
if (!isset($_SERVER['FRANKENPHP_WORKER'])) {
    handle_request();
    die();
}

// if we are in worker mode, we will handle ~1000 requests and then restart the worker,
// we add a random jitter to prevent all the workers from restarting at the same time.
$maxRequests = random_int(900,1100);
for ($numRequests = 0, $running = true; $numRequests < $maxRequests; ++$numRequests) {
    $running = frankenphp_handle_request(handle_request(...));

    // perform any worker maintenance here
    gc_collect_cycles();
}
```

When running `docker compose`, step debugging just works (most of the time, sometimes you have to edit the composer file with this env: `XDEBUG_CONFIG="client_host=<MY IP>"`) depending on network conditions/routing of the computer. For example, on wifi, I always have to add it, but when I'm jacked in with an ethernet connection, I don't have to add it. (:sigh: technology...)

**Reply by hazadam** [reply]
@withinboredom hi there, this sounds like it simply works out of the box for you and it does, for the most part, work for me too, but I'm fighting with the path mapping (Servers) configuration in PHPStorm since there's no "SERVER_NAME" value with the first dummy request after Frankens startup. How do you go about that? Have you not encountered any issues with that? My thread is [here](https://github.com/dunglas/frankenphp/discussions/495) btw. Thanks :slightly_smiling_face:

**Reply by withinboredom** [reply]
If you look carefully at the docker compose file, it disables worker mode for xdebug. My IDE (PHPStorm) gives a different error than what you are seeing but when I reported the bug to them, they closed as "won't fix".

I am uncertain how worker mode would actually work with xdebug since it replaces global variables, references, etc under the hood. It will probably segfault even if we did get it to work.

**Reply by hazadam** [reply]
Oh, bummer. We'll that's good to know. This information would've saved me many hours of debugging. A good docs candidate :laughing: . Thank you anyway :slightly_smiling_face:

**Reply by withinboredom** [reply]
> A good docs candidate

Indeed. However, I haven't needed worker mode + xdebug until very recently, and I've been using FrankenPHP since 2022 so it hasn't really come up.

**Reply by hazadam** [reply]
good news, I checked the source code once again and found that franken will pass env variables defined under given worker as $_SERVER variables. 
```
	frankenphp {
	    num_threads 2
		worker {
		    file /app/public/index.php
		    num 1
		    env SERVER_NAME mailship.local
		}
	}
``` 
Then after configuring xdebug to _start_with_request=yes_ and of coure correct _client_host_, PHPStorm would based on the newly added SERVER_NAME variable find path mapping and debugging now works as expected, that is **in worker mode**. :slightly_smiling_face:

**Reply by withinboredom** [reply]
Interesting! I'll have to give that a go as well!
